src/Interfaces.jl

```diff
@@ -94,22 +94,57 @@ function methodparts(IT, x::Expr)
     return methodname, args
 end
 
-requiredmethod(IT, nm, args, shouldthrow) = :(Interfaces.implemented($nm, $args, mods) || ($shouldthrow && Interfaces.missingmethod($IT, $nm, $args, mods)))
+"""
+Construct the method signature from a function name and argument types.
+
+    methodsig(:foo, Tuple{Int64, Float64, Any}) == :(foo(::Int64, ::Float64, ::Any))
+"""
+methodsig(f, args) = Expr(:call, f, unconvertargs(args)...)
+
+function requiredmethod(IT, nm, args, shouldthrow)
+    :(Interfaces.implemented($nm, $args, mods) || ($shouldthrow && Interfaces.missingmethod(debuglvl, $IT, $nm, $args, mods)))
+end
 
-function requiredreturn(T, nm, args, shouldthrow, RT_sym, __RT__)
+function requiredreturn(IT, nm, args, shouldthrow, RT_sym, __RT__)
     return quote
-        check = $(requiredmethod(T, nm, args, shouldthrow))
+        $(requiredmethod(IT, nm, args, shouldthrow))
         $RT_sym = Interfaces.returntype($nm, $args)
```

Comment:
I haven't thought through all the details here probably as much as you, but it seems like we'd probably want to add a check here if we're in `debug` mode and the return type is `NoMethod` to issue an error.

Reply:
yeah, we should only get `NoMethod` if we're in debug mode (because otherwise `requiredmethod` will throw an error if there isn't actually a method).

And if we do get `NoMethod`, then the next line will be `RT_sym <: NoMethod` which will always be `false`

So we just need to make sure that does correctly set `check = false`, and then `check || ($shouldthrow && ...)` will report the error

Reply:
I see. We could perhaps just skip calling `returntype` in debug mode if `check` is false, since we know it's going to return `NoMethod` anyway.

Reply:
yes, that sounds much nicer -- thanks!